tools/etcd-dump-db/backend.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2016 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/binary"
+	"fmt"
+	"path/filepath"
+
+	"github.com/boltdb/bolt"
+	"github.com/coreos/etcd/mvcc"
+	"github.com/coreos/etcd/mvcc/backend"
+)
+
+var (
+	keyBucketName          = []byte("key")
+	metaBucketName         = []byte("meta")
+	consistentIndexKeyName = []byte("consistent_index")
+)
+
+func snapDir(dataDir string) string {
+	return filepath.Join(dataDir, "member", "snap")
+}
+
+type revision struct {
+	main int64
+	sub  int64
+}
+
+func bytesToRev(bytes []byte) revision {
+	return revision{
+		main: int64(binary.BigEndian.Uint64(bytes[0:8])),
+		sub:  int64(binary.BigEndian.Uint64(bytes[9:])),
+	}
+}
+
+func listBucket(dbPath string) (buckets []string, err error) {
+	db, derr := bolt.Open(dbPath, 0600, &bolt.Options{})
+	if err != nil {
+		return nil, derr
+	}
+	defer db.Close()
+
+	err = db.View(func(tx *bolt.Tx) error {
+		return tx.ForEach(func(b []byte, _ *bolt.Bucket) error {
+			buckets = append(buckets, string(b))
+			return nil
+		})
+	})
+	return
+}
+
+func iterateBucket(dbPath, bucket string, limit uint64) (err error) {
+	db, derr := bolt.Open(dbPath, 0600, &bolt.Options{})
+	if err != nil {
+		return derr
+	}
+	defer db.Close()
+
+	err = db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(bucket))
+		if b == nil {
+			return fmt.Errorf("got nil bucket for %s", bucket)
+		}
+
+		c := b.Cursor()
+
+		// iterate in reverse order (use First() and Next() for ascending order)
+		var cnt uint64
+		for k, v := c.Last(); k != nil; k, v = c.Prev() {
+			fmt.Printf("key   = %s\n", string(k))
+			fmt.Printf("value = %s\n\n", string(v))
+
+			cnt++
+			if limit > 0 && cnt >= limit {
+				break
+			}
+		}
+
+		return nil
+	})
+	return
+}
+
+func getHash(dbPath string) (hash uint32, err error) {
+	db, derr := bolt.Open(dbPath, 0600, &bolt.Options{})
+	if err != nil {
+		return 0, derr
+	}
+	defer db.Close()
+
+	return backend.ComputeHash(db, mvcc.DefaultIgnores)
+}
+
+func revertDB(dbPath, copyPath string, revertRev int64) (err error) {
+	fmt.Println("copying", dbPath, "to", copyPath)
+	copy(dbPath, copyPath)
+
+	db, derr := bolt.Open(copyPath, 0600, &bolt.Options{})
+	if err != nil {
+		return derr
+	}
+	defer db.Close()
+
+	return revertByRevision(db, revertRev)
+}
+
+func revertByRevision(db *bolt.DB, revertRev int64) (err error) {
```

Comment:
this probably wont work anymore. the stresser test now might create leases which wont increase revision. how about removing this feature for now?